### PR TITLE
feat: payment detection for ethereum fee proxy 

### PR DESCRIPTION
**Pull request by vrolland**

### CODE REVIEW

File: packages/payment-detection/src/eth/fee-proxy-contract.ts [hunk 1]
```diff
@@ -0,0 +1,334 @@
+import { ethereumFeeProxyArtifact } from '@requestnetwork/smart-contracts';
```
Comment (benjlevesque) [inline packages/payment-detection/src/eth/fee-proxy-contract.ts:1]: this class is 90% identical to others... it's time for a refactoring don't you think?

File: packages/payment-detection/test/eth/proxy-info-retriever.test.ts [hunk 2]
```diff
@@ -22,7 +22,10 @@ describe('api/eth/proxy-info-retriever', () => {
       );
 
       // inject mock provider.getLogs()
-      infoRetriever.provider.getLogs = (): any => {
+      infoRetriever.provider.getLogs = (filter): any => {
+        if(filter.topics && filter.topics[0] === '0xa1c241e337c4610a9d0f881111e977e9dc8690c85fe2108897bb1483c66e6a96') {
```
Comment (benjlevesque) [inline packages/payment-detection/test/eth/proxy-info-retriever.test.ts:26]: this deserves an explanation.  couuld you add a comment ?

File: packages/payment-detection/src/eth/fee-proxy-contract.ts [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+/* eslint-disable max-classes-per-file */
+/** Exception when network not supported */
+class NetworkNotSupported extends Error {}
+/** Exception when version not supported */
+class VersionNotSupported extends Error {}
+
+/**
+ * Gets the payment proxy deployment information
+ */
+export type DeploymentInformationGetter = (
```
Comment (benjlevesque) [inline packages/payment-detection/src/eth/fee-proxy-contract.ts:28]: unused?

**Review by benjlevesque — CHANGES_REQUESTED** [review]

File: packages/payment-detection/src/eth/fee-proxy-contract.ts (hunk 1 above)
Comment (vrolland) [reply packages/payment-detection/src/eth/fee-proxy-contract.ts:1]: Discussed IRL. it will be my next priority. So let's keep it for a next PR.

**Review by benjlevesque — APPROVED** [review]

File: packages/payment-detection/src/eth/fee-proxy-contract.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import ProxyEthereumInfoRetriever from './proxy-info-retriever';
+import FeeReferenceBasedDetector from '../fee-reference-based-detector';
+
+// interface of the object indexing the proxy contract version
+interface IProxyContractVersion {
+  [version: string]: string;
+}
+
+const PROXY_CONTRACT_ADDRESS_MAP: IProxyContractVersion = {
+  ['0.2.0']: '0.2.0',
```
Comment (yomarion) [inline packages/payment-detection/src/eth/fee-proxy-contract.ts:13]: Why 0.2.0 ?

File: packages/payment-detection/src/eth/fee-proxy-contract.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+/**
+ * Handle payment networks with ETH input data extension
+ */
+export default class PaymentNetworkETHFeeProxy extends FeeReferenceBasedDetector<PaymentTypes.IETHPaymentEventParameters> {
```
Comment (yomarion) [inline packages/payment-detection/src/eth/fee-proxy-contract.ts:19]: Can you name it with something specific ? For example `ETHFeeProxyDetector`.

File: packages/payment-detection/src/eth/proxy-info-retriever.ts [hunk 6]
```diff
@@ -86,6 +110,9 @@ export default class ProxyEthereumInfoRetriever
         parameters: {
           block: blockNumber,
           txHash: transactionHash,
+          to: this.toAddress,
+          feeAddress: parsedLog.feeAddress || undefined,
```
Comment (yomarion) [inline packages/payment-detection/src/eth/proxy-info-retriever.ts:114]:
```suggestion
          feeAddress: parsedLog.feeAddress
```
Should be enough if parsedLog is always defined

**Review by yomarion — COMMENTED** [review]
Nice one ! My main concern is about the non-specific naming, "payment network" is not payment-detector specific enough, I feel. Same for the exported classes in `index.ts` by the way but I understand if you prefere to make it consistent and change everything at once in this one.

File: packages/payment-detection/src/eth/fee-proxy-contract.ts (hunk 4 above)
Comment (vrolland) [reply packages/payment-detection/src/eth/fee-proxy-contract.ts:13]: for bad reason, I fix it.

**Review by alexandre-abrioux — APPROVED** [review]

**Pull request merged**